### DISCUSSION TITLE
nlohmann::detail::type_error: [json.exception.type_error.302

### DISCUSSION THREAD

**Discussion by mohammedalasmar** [nlohmann/json#3510] [General]
This might be a simple question.

When I receive a json object and try to read a field in this object but that field doesn't exist, I get **nlohmann::detail::type_error: [json.exception.type_error.302]**

Is there any way to just get null or something like that instead of "terminating with uncaught exception"? Is there any error handler example available? 

Example:

```JS
// create json obj and send it
json jObj;
jObj["name"] = "nick";
..

// received  jObj
std::string n =  jObj["username"]; // username is not right so I get the terminating error
 ```

Many thanks! 
Mo

**Comment by nlohmann** [comment]
You can use `contains` (https://json.nlohmann.me/api/basic_json/contains/) to check if a key is present.

Also note a type error (302) is not thrown when a key is missing, but rather when you were able to retrieve a key, but the value type is not fitting the receiver type.

**Reply by mohammedalasmar** [reply]
Many thanks for the prompt reply! 
**`contains`** works for me. I also found **`is_null()`** which does what I wanted as well. 

 ```JS
 bool isFound = jObj.contains("name");
or
 bool isFound = jObj["name"].is_null();

 ```

**Reply by nlohmann** [reply]
Note the two options are not the same:

```cpp
jObj.contains("name");
```

will never change `jObj` and will work on const and non-const values.

```cpp
jObj["name"].is_null();
```

only returns `true`, because `operator[]` will insert a `null` value at key `name` in case there is no such key. When `jObj` is const, accessing a non-existing key is undefined behavior. See https://json.nlohmann.me/features/element_access/unchecked_access/ for more information.

tl;dr: The upper code is right. Don't use the lower one.

**Comment by tienviitta** [comment]
I got the same error but I want different behavior, e.g., how to handle `null` in JSON? For example I have this normally in my input JSON:

`
...
        "tv": {
            "disable": null,
            "enable": null,
            "on": false
        }
...
`

Those `disable` and `enable` are `std::vector<std::string>`s but those are specified only when `on` is `true`.

```
struct Tv {
    std::vector<std::string> disable;
    std::vector<std::string> enable;
    bool on;
};

void from_json(const nlo::json &j, Tv &tv) {
    j.at("disable").get_to(tv.disable);
    j.at("enable").get_to(tv.enable);
    j.at("on").get_to(tv.on);
}
```

So, should I check if the disable or enable is null with `.is_null()` in this case and if then set this as `nullptr`?

**Reply by tienviitta** [reply]
And this of course works when disable and enable are list of strings.

**Reply by tienviitta** [reply]
This would be my guess:

```
void from_json(const nlo::json &j, Tv &tv) {
    if (not j.at("disable").is_null())
        j.at("disable").get_to(tv.disable);
    if (not j.at("enable").is_null())
        j.at("enable").get_to(tv.enable);
    j.at("on").get_to(tv.on);
}
```
